stake-pool/program/src/processor.rs

```diff
@@ -1169,13 +1182,18 @@ impl Processor {
             return Err(StakePoolError::StakeLamportsNotEqualToMinimum.into());
         }
 
-        let current_minimum_delegation = minimum_delegation(stake_minimum_delegation);
-        if stake.delegation.stake > current_minimum_delegation {
+        // add the rent exempt reserve to the required delegation amount, since
+        // a decreasing stake must have at least rent-exemption plus minimum-delegation,
+        // and the source stake must still have minimum-delegation
+        let required_delegation = minimum_delegation(stake_minimum_delegation)
+            .saturating_mul(LAMPORT_BUFFER_FACTOR)
+            .saturating_add(meta.rent_exempt_reserve);
+        if stake.delegation.stake > required_delegation {
```

Comment:
> now required_lamports = 2(MINSTAKE + RENT) and required_delegation = 2MINSTAKE + RENT. multiplying the rent by two in the first calculation seems like a mistake, which makes the constraint looser by RENT than intended, in which case the two variables maybe should be equal

I thought so too at first, but in order to decrease a validator's stake, you need to be able to split a minimum stake from it, while maintaining minimum stake in the validator stake account. So that means you need to split `minimum_delegation + stake_rent`, while maintaining `minimum_delegation + stake_rent` in the source account, which means it needed `2 * (minimum_delegation + stake_rent)`

> but stepping back a bit... do we even need any of this? does anything bad happen if you remove a validator with extra stake or extra lamports in it? it all just goes back to the reserve either way, doesnt it?

This is an excellent point. The checks were from the times when the staker would put up the lamports for rent-exemption + minimum_delegation, so that the staker couldn't reclaim more than they put in. Now that everything comes from the reserve, there's no reason to keep this so complicated, and instead simplify the check to "if there's transient stake, make sure it's deactivating"